### ISSUE TITLE
Photon vs ORM - Schema Mapping

### ISSUE THREAD

**Issue by brillout** [prisma/photonjs#240]
I like many divergences Photon/Lift takes from classical ORMs. Most notably that Photon/Lift doesn't try to map classes with models.

Focusing on being *only* a tool to create simple & powerful queries seems like a sensible vision to me.

I have, however, questions about how Photon/Lift maps its schema with the database schema. I've skimmed through your blog posts and docs but didn't see any writings related to this.

How does Photon/Lift map its schema with the database schema?

It seems crucial to me that this mapping happens in a canonical, simple, and natural way. This would enabled me to use the full power of my database while using Photon/Lift.

When I choose MongoDB then I want to use all the niceness of MongoDB. Same for PostgreSQL.

For example, does Photon/Lift allow me to use my PostgreSQL database with [gajus/slonik](https://github.com/gajus/slonik)? Could I, hypothetically, progressively replace all Photon queries with SQL queries and eventually remove Photon/Lift from my project? (Not that this would make sense, but just to know that Photon/Lift will never get in the way of using the full power of PostgreSQL.)

In short: I don't want Photon to entirely abstract the database away from me. Instead, I want it to be a tool that I can use in parallel to the multitude of tools that my database offers.

If this is one of your *primary* goal, then what you guys are doing is super exciting to me.

**Comment by divyenduz** [comment]
This is indeed one of our primary goals. 

I suggest that you read these two specs to get an understanding of how schema and Prisma framework tools tie together: 

https://github.com/prisma/specs/tree/master/schema
https://github.com/prisma/specs/tree/master/lift

Additionally, there would be a raw API in future that would let you manually build SQL (albeit we will work on making it a good user experience). That being said, you can from any part of your application bypass Photon and write an SQL query directly.

**Comment by brillout** [comment]
>  the concept of capabilities

Capabilities are about making Primsa queries as powerful as possible, correct?

>  raw API spec

The raw API spec is about mapping the Primsa schema to the Database schema: so that I can write database queries (e.g. SQL) without having to know the column/table names of my database, correct?

**Comment by brillout** [comment]
Any luck to get clarification on this?

AFAICT the mapping between prisma schema and database schema is not canonical and seems intricate. That's a fundamental problem; I don't want to use Photon for everything. I want to use Photon for 90% of my queries and use my beloved native database queries for the rest. When I use database queries, I want Photon to go *entirely* out of my way. This isn't the case as things are.

I want to use Photon for simple CRUD operations but I do not want Photon to cripple my database and SQL skills.

Many devS are (rightfully) concerned about the ORM impedance mismatch. If you'd make the schema mapping trivial then you'd avoid the ORM impedance mismatch entirely. But I'm afraid you guys are not interested in that. Too bad. Am I wrong?

I really do want to love Photon but I need to keep my freedom and my database power.

cc
@timsuchanek 
@nikolasburk 
@divyenduz
@matthewmueller 
@schickling

**Comment by matthewmueller** [comment]
Hey @brillout! Thanks for taking the time to share your thoughts. I've responded inline:

> Capabilities are about making Prisma queries as powerful as possible, correct?

Capabilities are about making database access as type-safe as possible. Translating your cobbled-together SQL strings into a type-safe API in the language of your choice.

> The raw API spec is about mapping the Prisma schema to the Database schema: so that I can write database queries (e.g. SQL) without having to know the column/table names of my database, correct?

Yep, that's right. Maybe this wasn't clear from the spec, but there's a 1:1 mapping between model:table names and field:column names. The only reason we do this is to provide *minimal* type-safety over raw SQL strings.

> AFAICT the mapping between prisma schema and database schema is not canonical and seems intricate.

I totally agree with this issue and it's something I've been pushing for internally since day 1. Your database modeling skills should transfer smoothly to Prisma and you shouldn't have to learn the Prisma Way ™️.

I've been going through and trying to remove opinions from our Prisma schema:

- Make relationships explicit: https://github.com/prisma/specs/pull/299
- Use `prisma format` to speed this up: https://github.com/prisma/specs/issues/267

It's very early and not final, but please take a look and let me know if you have any thoughts.

> I want to use Photon for simple CRUD operations but I do not want Photon to cripple my database and SQL skills.

I think the previous replies answer this, but please let me know if I need to clarify!
